components/omega/doc/design/Metadata.md

```diff
@@ -0,0 +1,320 @@
+    Metadata(const std::string name); // name to assign
+
+    // construct metadata for a variable/field using
+    // required metadata. Not that if stdName doesn't exist
+    // an empty string can be supplied
+    Metadata(const std::string name,        // name of var
+             const std::string description, // long name or description
+             const std::string units,       // units
+             const std::string stdName,     // CF standard name
+             const int         nDims,       // number of dimensions
+             const std::vector<MetadataDims*> // vector of dim pointers
+             );
+
+Do we want to require any other metadata (valid min/max, fill value)?
+
+A destructor will be provided to free metadata and remove from the
+defined list, though we anticipate most metadata will need to be
+persistent through a given simulation.
+
+#### 4.2.2 Add/remove metadata
+
+An interface for adding new metadata to a previously defined Metadata
+instance will be provided. This can be used to add global metadata to
+the code, simulation, file metadata, but also for variables if additional
+metadata is desired. This will be an aliased function based on
+the data type of the metadata.
+
+    ierr = MetadataAdd(varName, name, value);
+
+where varName is the name of the already defined variable, and
+the other arguments are the usual name/value pair for the metadata
+to be added. This function will return 0 if successful and an error
+if either the varName has not been defined or if a metadata pair with
+that name already exists (or maybe we want to allow overwrite?).
+
+For symmetry, we will supply a remove function, though the
+use case is likely rare.
+
+    ierr = MetadataRemove(varName, name); 
+
+This doesn't need a type, though it might make it easier
+to search.
+
+#### 4.2.3 Retrieve metadata
+
+The most common use case will be retrieving metadata. For a single
+metadata entry, there will be an explicit get function:
+
+    myValue = MetadataGet(varName, name);
+
+where varName is the field defined (or generic name for global
+metadata) and name is the name associated with the metadata
+to be retrieved.
+
+During I/O stages, we will need a capability for retrieving
+all metadata for a defined field or global category. We will
+supply a function to retrieve the map associated with a given
+data type. The `std::map` iterators and functionality can then
+be used to extract all of the metadata for a given type. For
+example, to retrieve all R4 metadata from a defined field:
+
+   std::map<std::string, R4> r4meta = MetadataGetAll(varName);
+   std::string metaName;
+   R4 metaVal
+
+   // use std::map functionality to loop through all metadata
+   for (auto it=r4meta.begin(); it != r4.meta.end(), ++it){
+      metaName = it->first;  // retrieve name part of meta pair
+      metaVal  = it->second; // retrieve value part of meta pair
+
+      // do whatever needed with metadata
+   }
```

Comment:
Here in particular, the GitHub preview of this document isn't coming out quite right because you have 3 spaces rather than 4.  An easy way to avoid this is to put code blocks in triple-back-quotes.  If you specify `c++`:
````
```c++
```
````
 the code will be colored according to syntax for improved (at least in my view) readability.  Here, that would become:
````suggestion
```c++
std::map<std::string, R4> r4meta = MetadataGetAll(varName);
std::string metaName;
R4 metaVal

// use std::map functionality to loop through all metadata
for (auto it=r4meta.begin(); it != r4.meta.end(), ++it){
   metaName = it->first;  // retrieve name part of meta pair
   metaVal  = it->second; // retrieve value part of meta pair
    // do whatever needed with metadata
}
```
````